### ISSUE TITLE
epoll does not work on Raspberry Pi

### ISSUE THREAD

**Issue by operutka** [carllerche/nix-rust#181]
Hi,

I'm facing problems with epoll on Raspberry Pi. I've figured out it's due to different size of the EpollEvent structure. The structure defined in rust comes out packed but the native structure is not packed (i.e. it's 16 bytes wide). In my opinion, it's common pattern for the whole arm architecture, so the code in `sys/epoll.rs` should be changed to something like this:

```rust
#[cfg(target_arch = "arm")]
#[derive(Clone, Copy)]
#[repr(C)]
pub struct EpollEvent {
    ...
}

#[cfg(target_arch = "arm")]
#[test]
fn test_epoll_event_size() {
    ...
}

#[cfg(not(target_arch = "arm"))]
#[derive(Clone, Copy)]
#[repr(C, packed)]
pub struct EpollEvent {
    ...
}
```

I can send you a pull request if you want.

**Comment by carllerche** [comment]
Makes sense. I don't have a raspberry pi setup so I can't test. Is there a way to add it to CI?

**Comment by operutka** [comment]
I've already tested it and it works fine with RPi. Unfortunately, I cannot test it on other arm platforms. I'm not sure what you mean by CI but in case you mean Continuous Integration - it should be quite easy even though I don't know your CI setup. You just need a cross-compiler and some HW to test the binary.

Follow these steps to build your cross-compiler:
[https://github.com/npryce/rusty-pi/blob/master/doc/compile-the-compiler.asciidoc]

To make it working with cargo, you'll have to add these lines into your `~/.cargo/config`:

```toml
[target.arm-unknown-linux-gnueabihf]
ar     = "arm-linux-gnueabihf-ar"
linker = "arm-linux-gnueabihf-gcc"
```

Assuming you have `pi-tools` and `pi-rust` (i.e. the cross-compiling tools) in your home directory and that your build platform is x86\_64, add these into your PATH:

```
~/pi-tools/arm-bcm2708/gcc-linaro-arm-linux-gnueabihf-raspbian-x64/bin
~/pi-rust/bin
```

these into your LD_LIBRARY_PATH:

```
~/pi-tools/arm-bcm2708/gcc-linaro-arm-linux-gnueabihf-raspbian-x64/lib
~/pi-rust/lib
```

and export a new CC (this is required in case you use a cargo build script to build some native lib):

```
export CC=~/pi-tools/arm-bcm2708/gcc-linaro-arm-linux-gnueabihf-raspbian-x64/bin/arm-linux-gnueabihf-gcc
```

The last step is to pass `--target=arm-unknown-linux-gnueabihf` parameter when running cargo.

**Comment by carllerche** [comment]
I'll talk a look. I use travis. It would be nice to try to run the tests in a virtualized arm environment or something...

**Comment by carllerche** [comment]
Fixed by #c6cad77d10836f12ecda7bee5fb7a57ee6ea4f89
